price-feeder/oracle/client/chain_height.go

```diff
@@ -0,0 +1,113 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/rs/zerolog"
+	tmrpcclient "github.com/tendermint/tendermint/rpc/client"
+	tmctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+var (
+	errParseEventDataNewBlockHeader = errors.New("error parsing EventDataNewBlockHeader")
+	queryEventNewBlockHeader        = tmtypes.QueryForEvent(tmtypes.EventNewBlockHeader)
+)
+
+// ChainHeight is used to cache the chain height of the
+// current node which is being updated each time the
+// node sends an event of EventNewBlockHeader.
+type ChainHeight struct {
+	Logger zerolog.Logger
+
+	mtx               sync.RWMutex
+	errGetChainHeight error
+	lastChainHeight   int64
+}
+
+var (
+	// keep the instance thread safe.
+	once sync.Once
+
+	// keeps the single instance of ChainHeight.
+	instance *ChainHeight
+)
+
+// ChainHeightInstance returns the single instance of ChainHeight.
+func ChainHeightInstance(
+	ctx context.Context,
+	rpcClient tmrpcclient.Client,
+	logger zerolog.Logger,
+) (*ChainHeight, error) {
+	if !rpcClient.IsRunning() {
+		if err := rpcClient.Start(); err != nil {
+			return nil, err
+		}
+	}
+
+	newBlockHeaderSubscription, err := rpcClient.Subscribe(
+		ctx, tmtypes.EventNewBlockHeader, queryEventNewBlockHeader.String())
+	if err != nil {
+		return nil, err
+	}
+
+	once.Do(func() {
+		instance = &ChainHeight{
+			Logger:            logger.With().Str("oracle_client", "chain_height").Logger(),
+			errGetChainHeight: nil,
+			lastChainHeight:   0,
+		}
+
+		go instance.keepUpdating(ctx, rpcClient, newBlockHeaderSubscription)
+	})
+
+	return instance, nil
+}
+
+// UpdateChainHeight receives the data to be updated thread safe.
+func (chainHeight *ChainHeight) UpdateChainHeight(blockHeight int64, err error) {
```

Comment:
Let's change it to a private function.

Reply:
done at https://github.com/umee-network/umee/pull/1069/commits/1a19e3715ccf1a7641df54326edb5870d3b21a59